compiler/src/dotty/tools/debug/InsertExpression.scala

```diff
@@ -0,0 +1,238 @@
+        |    method.setAccessible(true)
+        |    val res = unwrapException(method.invoke(obj, args*))
+        |    if returnTypeName == "void" then () else res
+        |  }
+        |
+        |  def callConstructor(className: String, paramTypesNames: Array[String], args: Array[Object]): Any = {
+        |    val clazz = classLoader.loadClass(className).nn
+        |    val constructor = clazz.getConstructors.nn
+        |      .find { c => c.getParameterTypes.nn.map(_.nn.getName).toSeq == paramTypesNames.toSeq }
+        |      .getOrElse(throw new NoSuchMethodException(s"new $$className"))
+        |    constructor.setAccessible(true)
+        |    unwrapException(constructor.newInstance(args*))
+        |  }
+        |
+        |  def getField(obj: Any, className: String, fieldName: String): Any = {
+        |    val clazz = classLoader.loadClass(className).nn
+        |    val field = clazz.getDeclaredField(fieldName).nn
+        |    field.setAccessible(true)
+        |    field.get(obj)
+        |  }
+        |
+        |  def setField(obj: Any, className: String, fieldName: String, value: Any): Any = {
+        |    val clazz = classLoader.loadClass(className).nn
+        |    val field = clazz.getDeclaredField(fieldName).nn
+        |    field.setAccessible(true)
+        |    field.set(obj, value)
+        |  }
+        |
+        |  def getOuter(obj: Any, outerTypeName: String): Any = {
+        |    val clazz = obj.getClass
+        |    val field = getSuperclassIterator(clazz)
+        |      .flatMap(_.getDeclaredFields.nn.toSeq)
+        |      .map(_.nn)
+        |      .find { field => field.getName == "$$outer" && field.getType.nn.getName == outerTypeName }
+        |      .getOrElse(throw new NoSuchFieldException("$$outer"))
+        |    field.setAccessible(true)
+        |    field.get(obj)
+        |  }
+        |
+        |  def getStaticObject(className: String): Any = {
+        |    val clazz = classLoader.loadClass(className).nn
+        |    val field = clazz.getDeclaredField("MODULE$$").nn
+        |    field.setAccessible(true)
+        |    field.get(null)
+        |  }
+        |
+        |  def getSuperclassIterator(clazz: Class[?] | Null): Iterator[Class[?]] =
+        |    Iterator.iterate(clazz)(_.nn.getSuperclass).takeWhile(_ != null).map(_.nn)
+        |
+        |  // A fake method that is used as a placeholder in the extract-expression phase.
+        |  // The resolve-reflect-eval phase resolves it to a call of one of the other methods in this class.
+        |  def reflectEval(qualifier: Object, term: String, args: Array[Object]): Any = ???
+        |
+        |  private def unwrapException(f: => Any): Any =
+        |    try f catch {
+        |      case e: InvocationTargetException => throw e.getCause.nn
+        |    }
+        |
+        |  extension [T] (x: T | Null) {
+        |    private def nn: T = x.asInstanceOf[T]
+        |  }
```

Comment:
Don't we get this for free from `Predef`?

If the point is to protect against unimports of `Predef`, add `import scala.Predef.nn` at the beginning of the class instead?

Reply:
Done. Also I could remove almost all `.nn` thanks to flexible types. And I added a test with `-Yexplicit-nulls`.